File: python/model/pure_lstm.py

import os.path as path

import tensorflow as tf

from python.operator import lstm_aligned, connectivity
from python.model.abstact import AbstactModel

dirname = path.dirname(path.realpath(__file__))


class PureLSTM(AbstactModel):
    def __init__(self, dataset, name='pure_lstm',
                 embedding_size=1200, layers=2, **kwargs):
        super().__init__(
            dataset,
            params={
                'embedding_size': embedding_size,
                'layers': layers,
                'source_classes': dataset.source_classes,
                'target_classes': dataset.target_classes
                },
            name=name,
            **kwargs
        )

    def _model_fn(self, features, labels, mode, params):
        source = features['source']
        length = features['length']
        # If `mode == PREDICT` labels will be None, to make the graph
        # consistent for all models, just default it to an zero tensor.
        target = tf.zeros_like(source) if labels is None else labels

        # compute logits for model
        logits, embedding, embedding_matrix = lstm_aligned(
            source, length,
            source_vocab_size=params['source_classes'],
            target_vocab_size=params['target_classes'],
            latent_dim=params['embedding_size'],
            layers=params['layers'],
            target=features['target']
        )

        # compute prediction
        predict = tf.argmax(logits, 2)

        # compute losses
        loss = tf.losses.sparse_softmax_cross_entropy(
            logits=logits,
            labels=target,
            # mask <eos> <unknown> out of losses, (t != 0 && t != 1)
            weights=tf.logical_and(
                tf.not_equal(target, tf.zeros_like(target)),
                tf.not_equal(target, tf.ones_like(target))
            ),
            reduction=tf.losses.Reduction.MEAN
        )

        # Create optimizer
        optimizer = tf.train.AdamOptimizer()
        train_op = optimizer.minimize(
            loss,
            global_step=tf.train.get_global_step()
        )

        # Compute connectivity
        connectivity_sample = tf.zeros_like(source, dtype='float32')
        if 'offset' in features:
            connectivity_sample = connectivity(
                logits, features['target'],
                embedding, embedding_matrix,
                features['offset']
            )

        # Return EstimatorSpec, depending on the mode tensorflow will
        # automatically select the parameters that are required.
        return tf.estimator.EstimatorSpec(
            mode,
            loss=loss,
            train_op=train_op,
            predictions={
                'length': length,
                'source': source,
                'target': features['target'],

                'logits': logits,
                'probabilities': tf.nn.softmax(logits),
                'predict': predict,

                'connectivity': connectivity_sample
            },
            eval_metric_ops={
                'accuracy': tf.metrics.accuracy(labels=target,
                                                predictions=predict),
                'cross_entropy': tf.metrics.mean(loss)
            }
        )

File: article/models/pure_lstm.js

const dl = require('deeplearn');
const AbstactModel = require('./abstact_model.js');
const embeddingMatrix = require('../operator/embedding_matrix.js');
const BasicLSTMCell = require('../operator/basic_lstm_cell.js');
const MultiRNNCell = require('../operator/multi_rnn_cell.js');
const dynamicRnn = require('../operator/dynamic_rnn.js');
const denseLayer = require('../operator/dense_layer.js');

class PureLSTM extends AbstactModel {
  constructor(dataDirectory, name='autocomplete_lstm') {
    super({
      'embedding_size': 600,
      'layers': 2
    }, dataDirectory, name);
  }

  async _model_fn(features, variables, params) {
    const source = features.source;
    const length = features.length;

    // Create embedding matrix with zeros for <eos>
    const sourceEmbedding = embeddingMatrix(
      await variables.get('embedding-source')
    );

    // shape = [time, embedding_size]
    let logits = dl.gather(sourceEmbedding, source);

    // Setup RNN layers
    const rnnLayers = []
    for (let l = 0; l < params.layers; l++) {
      rnnLayers.push(new BasicLSTMCell(
        params['embedding_size'],
        dl.scalar(1),
        variables.scope(`rnn/multi_rnn_cell/cell_${l}/basic_lstm_cell`)
      ));
    }

    const multiRnnCell = new MultiRNNCell(rnnLayers);
    const [outputs, states] = await dynamicRnn(
      multiRnnCell,
      logits,
      source.shape[0]
    );

    // Get the output for the last layer for the last time step
    logits = outputs[outputs.length - 1];

    // Do a tf.layers.dense
    logits = await denseLayer(
      logits,
      variables.scope('dense')
    );

    // Compute predicted labels
    const predict = dl.argMax(logits, 1)

    // Compute softmax
    const probability = dl.softmax(logits)

    return [logits, probability]
  }
}

module.exports = PureLSTM;
